### PR TITLE
Backport the `UnwrapTwigExceptionListener`

### PR DESCRIPTION
**Pull request by bytehead**
Fixes a runtime error when using download elements e.g. in news in conjunction with twig templates.

The listener exists in Contao 5.x only.

```twig
# templates/news_full.html.twig
{{ text|raw }}
```

Bildschirmfoto 2024-01-24 um 21 55 30

```
[3/3] RuntimeError

Twig\Error\RuntimeError:
An exception has been thrown during the rendering of a template ("Error evaluating "text": This exception has no message. Use $exception->getResponse() instead.").

  at templates/news_full.html.twig:7
  at Twig\Template->displayWithErrorHandling(...)
     (vendor/twig/twig/src/Template.php:367)
  at Twig\Template->display(...)
     (vendor/twig/twig/src/Template.php:379)
  at Twig\Template->render(...)
     (vendor/twig/twig/src/TemplateWrapper.php:38)
  at Twig\TemplateWrapper->render(...)
     (vendor/twig/twig/src/Environment.php:280)
  at Twig\Environment->render('@Contao/news_full.html.twig', array(...))
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/TemplateInheritance.php:379)
  at Contao\Template->renderTwigSurrogateIfExists()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/TemplateInheritance.php:81)
  at Contao\Template->inherit()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Template.php:295)
  at Contao\Template->parse()
     (vendor/contao/core-bundle/src/Resources/contao/classes/FrontendTemplate.php:43)
  at Contao\FrontendTemplate->parse()
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNews.php:242)
  at Contao\ModuleNews->parseArticle(object(NewsModel))
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNewsReader.php:144)
  at Contao\ModuleNewsReader->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNewsReader.php:76)
  at Contao\ModuleNewsReader->generate()
     (vendor/contao/core-bundle/src/Resources/contao/elements/ContentModule.php:98)
  at Contao\ContentModule->generate()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:622)
  at Contao\Controller::getContentElement(object(ContentModel), 'main')
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:197)
  at Contao\ModuleArticle->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:70)
  at Contao\ModuleArticle->generate(false)
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:550)
  at Contao\Controller::getArticle(object(ArticleModel), false, false, 'main')
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:392)
  at Contao\Controller::getFrontendModule('0', 'main')
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:190)
  at Contao\PageRegular->prepare(object(PageModel))
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:60)
  at Contao\PageRegular->getResponse(object(PageModel), true)
     (vendor/contao/core-bundle/src/Resources/contao/controllers/FrontendIndex.php:320)
  at Contao\FrontendIndex->renderPage(object(PageModel))
     (vendor/symfony/http-kernel/HttpKernel.php:163)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:43)                
```

```
[2/3] RuntimeException 

RuntimeException:
Error evaluating "text": This exception has no message. Use $exception->getResponse() instead.

  at vendor/contao/core-bundle/src/Twig/Interop/ContextFactory.php:152
  at class@anonymous/vendor/contao/core-bundle/src/Twig/Interop/ContextFactory.php:119$3a4->__toString()
     (var/cache/dev/twig/3e/3e3670297c7e7c5fea354cb1ee65239f.php:51)
  at __TwigTemplate_bd45c2ea17c13d5c90acc91e1ca0cd98->doDisplay(...)
     (vendor/twig/twig/src/Template.php:394)
  at Twig\Template->displayWithErrorHandling(...)
     (vendor/twig/twig/src/Template.php:367)
  at Twig\Template->display(...)
     (vendor/twig/twig/src/Template.php:379)
  at Twig\Template->render(...)
     (vendor/twig/twig/src/TemplateWrapper.php:38)
  at Twig\TemplateWrapper->render(...)
     (vendor/twig/twig/src/Environment.php:280)
  at Twig\Environment->render('@Contao/news_full.html.twig', array(...))
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/TemplateInheritance.php:379)
  at Contao\Template->renderTwigSurrogateIfExists()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/TemplateInheritance.php:81)
  at Contao\Template->inherit()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Template.php:295)
  at Contao\Template->parse()
     (vendor/contao/core-bundle/src/Resources/contao/classes/FrontendTemplate.php:43)
  at Contao\FrontendTemplate->parse()
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNews.php:242)
  at Contao\ModuleNews->parseArticle(object(NewsModel))
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNewsReader.php:144)
  at Contao\ModuleNewsReader->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNewsReader.php:76)
  at Contao\ModuleNewsReader->generate()
     (vendor/contao/core-bundle/src/Resources/contao/elements/ContentModule.php:98)
  at Contao\ContentModule->generate()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:622)
  at Contao\Controller::getContentElement(object(ContentModel), 'main')
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:197)
  at Contao\ModuleArticle->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:70)
  at Contao\ModuleArticle->generate(false)
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:550)
  at Contao\Controller::getArticle(object(ArticleModel), false, false, 'main')
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:392)
  at Contao\Controller::getFrontendModule('0', 'main')
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:190)
  at Contao\PageRegular->prepare(object(PageModel))
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:60)
  at Contao\PageRegular->getResponse(object(PageModel), true)
     (vendor/contao/core-bundle/src/Resources/contao/controllers/FrontendIndex.php:320)
  at Contao\FrontendIndex->renderPage(object(PageModel))
     (vendor/symfony/http-kernel/HttpKernel.php:163)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:43)                
```

```
[1/3] ResponseException

Contao\CoreBundle\Exception\ResponseException:
This exception has no message. Use $exception->getResponse() instead.

  at vendor/contao/core-bundle/src/Resources/contao/library/Contao/File.php:796
  at Contao\File->sendToBrowser('', false)
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:1432)
  at Contao\Controller::sendFileToBrowser('files/downloads/foobar.pdf', false)
     (vendor/contao/core-bundle/src/Resources/contao/elements/ContentDownload.php:73)
  at Contao\ContentDownload->generate()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:622)
  at Contao\Controller::getContentElement(object(ContentModel))
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNews.php:128)
  at Contao\ModuleNews->Contao\{closure}()
     (vendor/contao/core-bundle/src/Twig/Interop/ContextFactory.php:141)
  at class@anonymous/vendor/contao/core-bundle/src/Twig/Interop/ContextFactory.php:119$3a4->__invoke()
     (vendor/contao/core-bundle/src/Twig/Interop/ContextFactory.php:150)
  at class@anonymous/vendor/contao/core-bundle/src/Twig/Interop/ContextFactory.php:119$3a4->__toString()
     (var/cache/dev/twig/3e/3e3670297c7e7c5fea354cb1ee65239f.php:51)
  at __TwigTemplate_bd45c2ea17c13d5c90acc91e1ca0cd98->doDisplay(...)
     (vendor/twig/twig/src/Template.php:394)
  at Twig\Template->displayWithErrorHandling(...)
     (vendor/twig/twig/src/Template.php:367)
  at Twig\Template->display(...)
     (vendor/twig/twig/src/Template.php:379)
  at Twig\Template->render(...)
     (vendor/twig/twig/src/TemplateWrapper.php:38)
  at Twig\TemplateWrapper->render(...)
     (vendor/twig/twig/src/Environment.php:280)
  at Twig\Environment->render('@Contao/news_full.html.twig', array(...))
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/TemplateInheritance.php:379)
  at Contao\Template->renderTwigSurrogateIfExists()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/TemplateInheritance.php:81)
  at Contao\Template->inherit()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Template.php:295)
  at Contao\Template->parse()
     (vendor/contao/core-bundle/src/Resources/contao/classes/FrontendTemplate.php:43)
  at Contao\FrontendTemplate->parse()
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNews.php:242)
  at Contao\ModuleNews->parseArticle(object(NewsModel))
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNewsReader.php:144)
  at Contao\ModuleNewsReader->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/news-bundle/src/Resources/contao/modules/ModuleNewsReader.php:76)
  at Contao\ModuleNewsReader->generate()
     (vendor/contao/core-bundle/src/Resources/contao/elements/ContentModule.php:98)
  at Contao\ContentModule->generate()
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:622)
  at Contao\Controller::getContentElement(object(ContentModel), 'main')
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:197)
  at Contao\ModuleArticle->compile()
     (vendor/contao/core-bundle/src/Resources/contao/modules/Module.php:214)
  at Contao\Module->generate()
     (vendor/contao/core-bundle/src/Resources/contao/modules/ModuleArticle.php:70)
  at Contao\ModuleArticle->generate(false)
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:550)
  at Contao\Controller::getArticle(object(ArticleModel), false, false, 'main')
     (vendor/contao/core-bundle/src/Resources/contao/library/Contao/Controller.php:392)
  at Contao\Controller::getFrontendModule('0', 'main')
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:190)
  at Contao\PageRegular->prepare(object(PageModel))
     (vendor/contao/core-bundle/src/Resources/contao/pages/PageRegular.php:60)
  at Contao\PageRegular->getResponse(object(PageModel), true)
     (vendor/contao/core-bundle/src/Resources/contao/controllers/FrontendIndex.php:320)
  at Contao\FrontendIndex->renderPage(object(PageModel))
     (vendor/symfony/http-kernel/HttpKernel.php:163)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:43)                
```

### CODE REVIEW

File: core-bundle/src/EventListener/UnwrapTwigExceptionListener.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+        if (!$throwable instanceof RuntimeError) {
+            return;
+        }
+
+        $previous = $throwable->getPrevious();
+
+        while ($previous && null !== $previous->getPrevious()) {
+            $previous = $previous->getPrevious();
+        }
```
Comment (m-vo) [inline core-bundle/src/EventListener/UnwrapTwigExceptionListener.php]: Why was this necessary again? Because we don't have these lines in the upstream version.

Comment (bytehead) [reply core-bundle/src/EventListener/UnwrapTwigExceptionListener.php]: I know. It's my additional code because in my case the exception is wrapped one level more (see stacktrace).

Comment (m-vo) [reply core-bundle/src/EventListener/UnwrapTwigExceptionListener.php]: How did that happen? If that's a typical scenario we would need the same in Contao 5, no?

Comment (bytehead) [reply core-bundle/src/EventListener/UnwrapTwigExceptionListener.php]: Hard to tell why it's wrapped once more. I could show you the scenario in a shared screen session :)

**Comment by bytehead** [comment]
Reason for `twig/twig ^3.8` is here: https://github.com/contao/contao/pull/6579#issuecomment-1833570625

**Review by m-vo — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bytehead.
